src/integrations/buckaroo/class-buckaroo.php

```diff
@@ -0,0 +1,571 @@
+
+    error_log(print_r($post_array, true));
+
+    return $post_array;
+  }
+
+  /**
+   * Make sure we have the data we need defined as filters.
+   *
+   * @throws \Missing_Filter_Info_Exception When not all required keys are set.
+   *
+   * @return void
+   */
+  protected function verify_buckaroo_info_exists(): void {
+    if ( empty( \apply_filters( Filters::BUCKAROO, 'website_key' ) ) ) {
+      throw Missing_Filter_Info_Exception::view_exception( Filters::BUCKAROO, 'website_key' );
+    }
+
+    if ( empty( \apply_filters( Filters::BUCKAROO, 'secret_key' ) ) ) {
+      throw Missing_Filter_Info_Exception::view_exception( Filters::BUCKAROO, 'secret_key' );
+    }
+  }
+
+  /**
+   * Get the correct url depending on if we're testing or not.
+   *
+   * @return string
+   */
+  protected function get_buckaroo_uri(): string {
+    return $this->is_test() ? $this->get_buckaroo_uri_test() : $this->get_buckaroo_uri_live();
+  }
+
+  /**
+   * Returns correct Buckaroo live uri.
+   *
+   * @return string
+   */
+  protected function get_buckaroo_uri_live(): string {
+    return $this->is_data_request() ? self::LIVE_URI_DATA_REQUEST : self::LIVE_URI_TRANSACTION;
+  }
+
+  /**
+   * Returns correct Buckaroo test uri.
+   *
+   * @return string
+   */
+  protected function get_buckaroo_uri_test(): string {
+    return $this->is_data_request() ? self::TEST_URI_DATA_REQUEST : self::TEST_URI_TRANSACTION;
+  }
+
+  /**
+   * Check if we're running a test or not.
+   *
+   * @return boolean
+   */
+  protected function is_test(): bool {
+    return $this->is_test_uri;
+  }
+
+  /**
+   * Check if we're running a test or not.
+   *
+   * @return boolean
+   */
+  protected function is_data_request(): bool {
+    return $this->is_data_request;
+  }
+
+  /**
+   * Set if you need to use the test URI instead of live one.
+   *
+   * @return void
+   */
+  public function set_test(): void {
+    $this->is_test_uri = true;
+  }
+
+  /**
+   * Set if you need to use the test URI instead of live one.
+   *
+   * @return void
+   */
+  public function set_data_request(): void {
+    $this->is_data_request = true;
+  }
+
+  /**
+   * Set's currency as uppercase 3-letter string (example: EUR)
+   *
+   * @param  string $currency Currency string.
+   * @return void
+   */
+  public function set_currency( string $currency ): void {
+    $this->currency = $currency;
+  }
+
+  /**
+   * Getter for $this->currency.
+   *
+   * @return string
+   */
+  public function get_currency(): string {
+    return $this->currency;
+  }
+
+  /**
+   * Get type of payment. Defaults to iDEAL.
+   *
+   * @return  string
+   */
+  public function get_pay_type() {
+     return $this->pay_type;
+  }
+
+  /**
+   * Set type of payment. Defaults to iDEAL.
+   *
+   * @param  string $pay_type  Type of payment. Defaults to iDEAL.
+   *
+   * @return  self
+   */
+  public function set_pay_type( string $pay_type ) {
+    $this->pay_type = $pay_type;
+    return $this;
```

Comment:
So the `set_pay_type()` returns the entire class? Why?

Reply:
Yeah not sure, I use an [VSCode extension for generating setters / getters](https://github.com/phproberto/vscode-php-getters-setters) and this is how it generates them, just didn't bother changing it. 🤷‍♂️  I would imagine this is so that you can chain setter calls like so:

```php
$class->set_something('something')->set_something_else('something else')...;
```

I can remove this if you like, it's not really used.

Reply:
I mean, I don't think it adds some super extra overhead so it can stay. I'm just not used to see it in WP I guess (I've seen this in Laravel tho).